### ISSUE TITLE
Return route responses in protobuf format

### ISSUE THREAD

**Issue by chaupow** [valhalla/valhalla#1348]
There's been [previous work](https://github.com/valhalla/valhalla/blob/master/src/tyr/route_serializer.cc#L2290) to return route responses in protobuf, but it [never made it over the finish line](https://github.com/valhalla/valhalla/blob/master/src/tyr/route_serializer.cc#L2274-L2288) to be used in an API call.

We should get this finished. 

Also, the former approach is transforming the response in this way:
- `C++ route response` -> `JSON` -> `Proto`. 

Instead I think we should implement this doing

- `C++ route response` -> `Proto` 

to reduce parsing work.

## Tasks

- [ ] Turn responses to [route proto objects](https://github.com/valhalla/valhalla/blob/master/proto/route.proto)
- [ ] optionally: Expose proto objects as a request param

**Comment by kevinkreiser** [comment]
The original idea is that the things that doing create are pbf  and we just return them directly without any transformation. To do this we wanted to basically refactor all the ones to nest properly into one meta object

**Comment by kevinkreiser** [comment]
Sorry for the originally totally borked comment the phone keyboard is garbage

**Comment by chaupow** [comment]
Cool. Thinking out loud here. 

If I want protobuf instead of json, I can modify `DirectionsBuilder::PopulateTripDirections` here 
https://github.com/valhalla/valhalla/blob/feature/pbf_response/src/odin/directionsbuilder.cc#L131 

And then populate the final pbf directly 
https://github.com/valhalla/valhalla/blob/feature/pbf_response/src/odin/directionsbuilder.cc#L131

**Comment by kevinkreiser** [comment]
@chaupow What I had planned to do was always fill out the pbf object (as valhalla does now) and then if the person wanted json, then I'd convert the pbf to json as a final step. If the person wanted pbf then its just `return pbf_obj.SerializeAsString();`.

As far as where to populate it, its a bit tricky. Right now we populate some of the request options in loki, then we make `n` TripPaths` in thor, and finally we make `TripDirections` in odin, then we serialize something from all of them in tyr.

If i were refactoring this I would make one object in pbf that encompasses all the things and then fill each part out along the way. In most cases we will not want to return the trip path, so we'll probably delete that off before serializing the pbf.

I hope this gives you some ideas!

**Comment by dgearhart** [comment]
@chaupow originally I was thinking that the `directions` proto would replace populating the TripDirections

**Comment by chaupow** [comment]
> @chaupow originally I was thinking that the directions proto would replace populating the TripDirections

Yes me too. This is a slightly/very different scope from [@kevinkreiser's proposal](https://github.com/valhalla/valhalla/issues/1348#issuecomment-394897952) even though that solution does not sound wrong. I'll keep digging and do whichever i feel suitable.

**Comment by devShahriar** [comment]
is there any workaround .
or is it implemented yet ?

I want send direct request with proto payload as and get response in proto .

As i am running some loadtest on valahalla to use it in the production I am getting 600-700 ms average ping .
and 76 throughput. 

I think if i send the payload in proto the response time will decrease .

Is it implemented ? or can anyone guide me how to achieve this

**Comment by kevinkreiser** [comment]
it is not implemented. i started a branch but have not yet pushed it. my idea was to both support this mode over http but also to support grpc.

so far i got to upgrading most of the fields to be able to work with protobuf 3. but that is all ive done so far.

when you say guide you to achieve it, do you mean you would like to implement it? there is a cheap way to start doing this with small number of changes but to me this seems like something that should be clearly marked as an unstable api. to do that we need only to add a new response format pbf, and then also start checkign the http content header to determine whether the postbody is json or pbf. this is not too much work imho and i could guide you to doing that if you are willing. would you like me to write this up are you willing to work on it?
